airbyte-lib: Track streams in cache

flash1293 opened the pull request:
This PR introduces an internal table to track which streams got synced to a cache.

This allows to read from a cache even if no source is available:
```
cache = ab.new_local_cache("cache_test")

# read from cache without a source
print(list(cache["stream1"]))
```

This can be used for situations where data is written to a cache in one script, reading from it in another.

Conceptually, the cache remembers what it stores in a persistent way not tied to python heap memory.

## Behavior

The cache already kept a `source_catalog` around with all streams that was read from. This information is now saved to an internal table when changed and loaded from the table when the cache object is created. New streams are merged into the existing catalog by the stream name, replacing existing stream definitions in case of collisions.

The internal table `_airbytelib_streams` is added to the same schema as all other tables.

When accessing streams from the cache object, all stream names that are in the catalog and also have an existing final table are available.

When accessing streams for a `ReadResult`, only the streams that got synced in that read are available, not all of them.

## Implementation details

As it's available already, this PR is using sqlalchemy ORM capabilities to create the table in the cache and manage its data.

## Unclear points

### Table name vs. stream name

A lot of the logic in the cache class is hinging on the stream name. However, the mapping from stream name to table name depends on the configuration (e.g. via the configured prefix). This leaves an edge case that could lead to confusing behavior:
* A cache is initialized without table name prefix and some data is synced
* Later in a different script, a new cache instance pointing to the same duckdb file is initialized which has a table name prefix configured
* The cache appears empty as for none of the streams in the internal streams table, a data table can be found

This could be "fixed" by leveraging the table name column in the internal streams table, but it makes the world more complex than justified IMHO.

### Different sources with identical stream names

As the interface of the cache is built around the stream name, there's no straightforward way to support multiple sources with identical stream names. I think it's OK to not handle this situation, just wanted to bring it up.

aaronsteers commented on airbyte-lib/airbyte_lib/results.py:
```diff
@@ -15,28 +15,39 @@
+        return stream in self._cache and stream in self._processed_streams
 
     def __iter__(self) -> Iterator[str]:
-        return self._cache.__iter__()
+        return [
+            stream_name
+            for stream_name in self._cache.__iter__()
+            if stream_name in self._processed_streams
+        ].__iter__()
```
Since I think this just needs the strings and not the dataset object, can we just iterate over `self._processed_streams` without referencing the cache?

flash1293 replied:
theoretically yes, let's go with that for now.

aaronsteers commented on airbyte-lib/airbyte_lib/results.py:
```diff
@@ -15,28 +15,39 @@
             raise KeyError(stream)
 
         return CachedDataset(self._cache, stream)
 
     def __contains__(self, stream: str) -> bool:
-        return stream in self._cache
+        return stream in self._cache and stream in self._processed_streams
```
Should we just check if it is in self._processed_streams? It seems that anything that is in processed streams list should also be in the cache (in theory at least).

aaronsteers commented on airbyte-lib/airbyte_lib/results.py:
```diff
@@ -15,28 +15,39 @@
 
     @property
     def streams(self) -> Mapping[str, CachedDataset]:
-        return self._cache.streams
+        return {
+            stream_name: CachedDataset(self._cache, stream_name)
+            for stream_name in self._cache.__iter__()
+            if stream_name in self._processed_streams
+        }
```
👍

aaronsteers commented on airbyte-lib/airbyte_lib/source.py:
```diff
@@ -369,10 +369,13 @@ def read(self, cache: SQLCacheBase | None = None) -> ReadResult:
             cache = get_default_cache()
 
-        cache.register_source(source_name=self.name, source_catalog=self.configured_catalog)
+        cache.register_source(
+            source_name=self.name, incoming_source_catalog=self.configured_catalog
+        )
```
Because sources can have a hundred or more streams, wdyt of passing the list of _selected_ streams (or all/None if nothing is selected). Since this registration occurs during "read()", we should know here which of the streams are actually going to be used and which will be ignored.

If we can implement without too much added complexity, I'd rather not store metadata on streams that are not going to receive data.

flash1293 replied:
This is already what happens - `configured_catalog` is filtered by the selected streams of the source: https://github.com/airbytehq/airbyte/blob/1268c3740b565aa773cf1643fbc98abc4d63c4a4/airbyte-lib/airbyte_lib/source.py#L177

aaronsteers replied:
Nice! Thanks for confirming.

aaronsteers commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -763,6 +792,93 @@ def _table_exists(
+        engine = self.get_sql_engine()
+        Base.metadata.create_all(engine)
+
+    @overrides
+    def register_source(
+        self,
+        source_name: str,
+        incoming_source_catalog: ConfiguredAirbyteCatalog,
+    ) -> None:
```
Noted also where this is called (quoted comment text below): I think it is worthwhile to filter the records we store in the metadata table down to those that we expect to receive data. One way to do that is to accept an optional stream_names filter in this method signature.

> Because sources can have a hundred or more streams, wdyt of passing the list of _selected_ streams (or all/None if nothing is selected). Since this registration occurs during "read()", we should know here which of the streams are actually going to be used and which will be ignored.
> 
> If we can implement without too much added complexity, I'd rather not store metadata on streams that are not going to receive data.
> 

 Wdyt?

flash1293 replied:
Linked the place where this is happening in the other comment.

aaronsteers commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -35,15 +46,27 @@
+class CachedStream(Base):  # type: ignore[valid-type,misc]
+    __tablename__ = STREAMS_TABLE_NAME
+
+    stream_name = Column(String, primary_key=True)
+    source_name = Column(String)
+    table_name = Column(String)
+    catalog_metadata = Column(String)
```
Wdyt of using a composite three-column primary key here?

```suggestion
    stream_name = Column(String, primary_key=True)
    source_name = Column(String, primary_key=True)
    table_name = Column(String, primary_key=True)
    catalog_metadata = Column(String)
```

The implications would be:

1. Dupes are allowed for different sources using the same stream name.
1. Dupes are allowed for the same stream name and some source name - if table suffix and/or suffix are different.
1. If a user configures the cache with different prefix/suffix settings, they'll basically be writing to a different namespace within the cache.
   1. Meaning, I could initialize my cache with "_test1" suffix, "_test2" suffix, etc., AirbyteLib would treat these as different namespaces with different tablesets.
1. When we get to supporting multiple sources in the same cache, we might want a dynamic default table prefix like `'{source_name}_'`. But we can tackle that later.

flash1293 replied:
The problem I'm seeing with that is that the cache api is stream name only based.

For example, you have the following entries in the _airbytelib_streams table:
| stream    | source  | table                      |
|-----------|---------|----------------------------|
| my_stream | source1 | source1_my_stream          |
| my_stream | source1 | source1_prefixed_my_stream |


If you do `cache["my_stream"]`, then what table are you getting? For this prefix case, we could decide based on the table name, by comparing the table name the cache would choose for a stream name with the table name stored in the _airbytelib_streams table. I'm going to implement this.

But what about this case:

| stream    | source  | table                      |
|-----------|---------|----------------------------|
| my_stream | source1 | source1_my_stream          |
| my_stream | source2 | source2_my_stream          |

 `cache["my_stream"]` could be either of those, and there is no way to decide. Changing this to  `cache["source1_my_stream"]` raises lots of new questions - it just doesn't feel like the right approach to me, this edge case shouldn't make it harder for the "regular" case.

flash1293 replied:
So, in summary, we should actually use the table name as primary key, because that's what needs to be unique for the current feature set.

For multi-source, I think it's OK to merge the streams and overwrite / raise in the collision case - after all, this is also what would happen in case of regular destinations.

aaronsteers replied:
Nice! Yeah, I think this sounds like a great path forward. 👍

aaronsteers reviewed (approved):
It's looking great. Some feedback inline 👍

flash1293 commented:
@aaronsteers I adjusted the PR as discussed above. This is the behavior now:
* Prefixes will serve as "namespaces" for cache (all maintained in the same internal table, but the user only "sees" the tables of the current namespace)
* If different sources write the same stream name, the last one will win (like in regular Airbyte destinations) - this allows us to keep the "stream name only" semantics when interacting with the cache which I think is something we should keep
* Moved the logic into a "catalog manager"

aaronsteers commented on airbyte-lib/airbyte_lib/caches/_catalog_manager.py:
```diff
@@ -0,0 +1,155 @@
+class CachedStream(Base):  # type: ignore[valid-type,misc]
+    __tablename__ = STREAMS_TABLE_NAME
+
+    stream_name = Column(String)
+    source_name = Column(String)
+    table_name = Column(String, primary_key=True)
+    catalog_metadata = Column(String)
```
I like this approach to the primary key. 👍👍

aaronsteers reviewed (approved):
Looks great. Thanks for the comments and related updates. Feel free to merge when ready.
